File: Code/app/src/main/java/com/bigfriedicecream/overcooked/presenters/RecipePresenter.java
package com.bigfriedicecream.overcooked.presenters;

import com.bigfriedicecream.overcooked.interfaces.IRecipeContract;
import com.bigfriedicecream.overcooked.models.RecipeDataModel;
import com.bigfriedicecream.overcooked.observables.RecipeRepository;

import java.util.Observable;
import java.util.Observer;

public class RecipePresenter implements IRecipeContract.Presenter, Observer {

    private IRecipeContract.View view;
    private RecipeRepository recipeRepository;

    public RecipePresenter(IRecipeContract.View v) {
        view = v;
        recipeRepository = RecipeRepository.getInstance();
    }

    public void start() {
        recipeRepository.addObserver(this);
    }

    public void stop() {
        recipeRepository.deleteObserver(this);
    }

    public void load(String id) {
        RecipeDataModel recipe = recipeRepository.get(id);
        if (recipe != null) {
            view.render(recipe);
        }
    }

    @Override
    public void update(Observable observable, Object o) {

    }
}

File: Code/app/src/main/java/com/bigfriedicecream/overcooked/presenters/RecipePresenter.kt
package com.bigfriedicecream.overcooked.presenters

import android.content.Context
import android.preference.PreferenceManager

import com.bigfriedicecream.overcooked.interfaces.IRecipeContract
import com.bigfriedicecream.overcooked.models.RecipeModel
import com.bigfriedicecream.overcooked.observables.RecipeRepository

import java.util.Observable
import java.util.Observer


class RecipePresenter(private val view:IRecipeContract.View) : IRecipeContract.Presenter, Observer {
    private val recipeRepository:RecipeRepository = RecipeRepository.instance
    private var recipeModel:RecipeModel = RecipeModel()
    private var id:String? = null
    private var adjustedQuantity:Int = -1

    override fun start(context:Context?, id:String?, restore:Boolean?) {
        this.id = id

        val preferences = PreferenceManager.getDefaultSharedPreferences(context)

        if (restore == true) {
            adjustedQuantity = preferences.getInt("adjustedQuantity", -1)

        } else {
            val preferenceEditor = preferences.edit()
            preferenceEditor.putInt("adjustedQuantity", -1)
            preferenceEditor.apply()
        }

        recipeRepository.addObserver(this)
        recipeRepository.load(context)
    }

    override fun stop() {
        recipeRepository.deleteObserver(this)
    }

    override fun adjustQuantityClick() {
        view.showAdjustQuantityDialog(recipeModel)
    }

    override fun updateQuantity(newQuantity:Int, context:Context?) {
        adjustedQuantity = newQuantity
        val preferences = PreferenceManager.getDefaultSharedPreferences(context).edit()
        preferences.putInt("adjustedQuantity", adjustedQuantity)
        preferences.apply()

        val recipeQuantity:Int = if (recipeModel.serves > 0) recipeModel.serves else recipeModel.makes
        val multiplier:Double = adjustedQuantity.toDouble() / recipeQuantity.toDouble()
        recipeModel.multiplier = multiplier
        view.render(recipeModel)
    }

    override fun update(observable:Observable, o:Any?) {
        if (observable is RecipeRepository) {
            recipeModel = recipeRepository.getRecipeById(id)

            if (adjustedQuantity != -1) {
                val recipeQuantity:Int = if (recipeModel.serves > 0) recipeModel.serves else recipeModel.makes
                val multiplier:Double = adjustedQuantity.toDouble() / recipeQuantity.toDouble()
                recipeModel.multiplier = multiplier
            }

            view.render(recipeModel)
        }
    }
}
